python/tvm/utils/__init__.py

```diff
@@ -0,0 +1,303 @@
+        }
+    )
+    with transform.PassContext(opt_level=3):
+        f = build(specialized, target=target)
+    a = nd.array(np.ones(vec_width * num_vector_registers * nthreads).astype("float32"), device=dev)
+    times = f.time_evaluator(f.entry_name, dev, repeat=100)(a)
+    flops = 2 * vec_width * num_vector_registers * nthreads * iters  # fma is two flops
+    flop_s = flops / times.min
+    return flop_s
+
+
+@T.prim_func
+def peak_bandwidth_tir(a: T.handle, b: T.handle, nt: T.int32, vec_width: T.int32) -> None:
+    # pylint: disable=invalid-name, missing-function-docstring
+    N = T.var("int32")
+    A = T.match_buffer(a, [N], "float32")
+    B = T.match_buffer(b, [nt * vec_width * 4], "float32")
+    # assert N % (nt * 4 * vec_width) == 0, "div"
+    # Parallelism is necessary to hit all cores/nodes
+    for i in T.parallel(nt):
+        for k in T.serial(N // nt // 4 // vec_width):
+            for l in T.unroll(4):
+                # vectorized load is necessary to hit peak bandwidth
+                for j in T.vectorized(vec_width):
+                    B[i * vec_width * 4 + l * vec_width + j] += A[
+                        i * (N // nt) + k * vec_width * 4 + l * vec_width + j
+                    ]
+
+
+def estimate_peak_bandwidth(target: Target, dev: Device, vec_width: Optional[int] = None) -> float:
+    """Estimate peak memory bandwidth of a target/device combo.
+
+    Peak bandwidth is estimated by running a small experiment on the underlying
+    hardware. The peak bandwidth measurement assumes that vector instructions
+    are being used to load the data.
+
+    Parameters
+    ----------
+    target : Target
+        Target to use for measurement. This target should be as specific to the
+        underlying hardware as possible.
+    dev : Device
+        Device to measure peak bandwidth on.
+    vec_width : Optional[int]
+        Vector unit width, determined from target if not supplied.
+
+    Returns
+    -------
+    float
+        Peak memory bandwidth in bytes/seconds.
+    """
+    # Ideally we'd be able to use this code to measure peak bandwidth of the
+    # different cache levels. If we could just generate load commands, then we
+    # could use those in a tight loop. Instead we need some code that is
+    # limited on the cache bandwidth. With the L1 cache we need an operation
+    # that has a very low arithmetic intensity and we haven't come up with one
+    # yet.
+    vec_width, _ = _vec_width_registers(target, vec_width, 1)
+    specialized = peak_bandwidth_tir.specialize(
+        {
+            peak_bandwidth_tir.params[3]: vec_width,
+        }
+    )
+    with transform.PassContext(opt_level=3):
+        f = build(specialized, target=target)
+    # Data size needs to be larger than last level of cache. We don't have a
+    # way of getting cache sizes, so this number should give us a large enough
+    # size.
+    size = 10**8 // (4 * num_threads() * vec_width) * (4 * num_threads() * vec_width)
```

Comment:
Do you mean execution order between threads? This assumes that threads are executing at the same time.

Ideally we would have a way to make sure that each load is directly from main memory, but there is no cross-platform/cross-architecture was of doing that (that I know of). If we make the data size large enough, then we can assume that the vast majority of loads don't hit cache. 2x LLC size would probably be a safe default, but it does depend a bit on the caching strategy.

Reply:
Hmmm, I see. Would it be worth it to expose size to the user? 

Also easy way to graph size vs estimated bandwidth and we can see when we hit large enough to avoid cache hit

Reply:
I'm not sure it is worth exposing the size to the user, then they have to go to more work to get the correct size. We could keep increasing the size until performance flatlines, but it would be computationally expensive and we would need a different parameter to control when to stop.

Reply:
It would be neat to do all the STREAM benches for the memory bandwidth measurement and combine them.

Reply:
Would also be interesting to apply the specific stream benchmark that applies to a given workload pattern based on the read and writes defined in a block. Probably diminishing returns given that they are usually comparable measurements of the bandwidth.

Reply:
Personally I think it would be nice to expose an interface to change this (you can set it default 10 ** 8) in the worst case for what csullivan says above

Reply:
> I'm just ramping into the discussion here but given cache size variability I would argue for the iterative approach for increasing the array size until performance plateaus. It's easy to measure and less likely to lead to false assumptions being made.

I would like to have a more robust approach, but the iterative approach also suffers from needing some sort of configuration. How do we decide that we have plateaued at memory bandwidth? What if we plateaued at LLC bandwidth? We'd have to run past a certain size to make sure it wasn't LLC, so we just end up running something around the size I've set here. The best solution is to have some way of getting LLC size and then doing a multiple of it, but we don't have a way to do that right now. Do you have a proposal for how I might determine if the bandwidth has plateaued or not?

> It would be neat to do all the STREAM benches for the memory bandwidth measurement and combine them.

I'm not sure the STREAM benchmarks are a good fit here because they involve writing back to memory, which may not be a reflection on what is happening in a regular ML kernel. And how would we combine them? Just average? I've looked at what other tools do to estimate bandwidth/flops and most have handwritten assembly or optimized kernels (like here) that just do load or just do flops.

> Would also be interesting to apply the specific stream benchmark that applies to a given workload pattern based on the read and writes defined in a block. Probably diminishing returns given that they are usually comparable measurements of the bandwidth.

I've thought about this idea, especially with regards to cache size/type of flops, but I haven't come up with a good way to determine cache sizes (for access patterns) or type of compute (FMA or not). It is definitely something I want to do, but I was trying to get a small working PR in first and then we can make the analysis more specific.

> Personally I think it would be nice to expose an interface to change this (you can set it default 10 ** 8) in the worst case for what csullivan says above

I guess we can expose it, but how would the user know what to set it to?

Reply:
Agreed there isn't a great way to reliably determine plateauing. I suppose could consider letting the user define a convergence function and then use something naive as a default (e.g. walk up the size in continuous steps and after a threshold of comparable runs, do a few additional steps of much larger interval?). 

But I think I'm okay with this as is for now, more desirably would be for this also to be a queryable attribute via the target attribute preprocessor flow.